extensions/panache/hibernate-reactive-panache-common/runtime/src/main/java/io/quarkus/hibernate/reactive/panache/common/runtime/CommonPanacheQueryImpl.java

```diff
@@ -0,0 +1,382 @@
+        }
+
+        // We use the first constructor that we found and use the parameter names,
+        // so the projection class must have only one constructor,
+        // and the application must be built with parameter names.
+        // Maybe this should be improved some days ...
+        Constructor<?> constructor = type.getDeclaredConstructors()[0];
+
+        // build select clause with a constructor expression
+        StringBuilder select = new StringBuilder("SELECT new ").append(type.getName()).append(" (");
+        int selectInitialLength = select.length();
+        for (Parameter parameter : constructor.getParameters()) {
+            if (!parameter.isNamePresent()) {
+                throw new PanacheQueryException(
+                        "Your application must be built with parameter names, this should be the default if" +
+                                " using Quarkus artifacts. Check the maven or gradle compiler configuration to include '-parameters'.");
+            }
+
+            if (select.length() > selectInitialLength) {
+                select.append(", ");
+            }
+            select.append(parameter.getName());
+        }
+        select.append(") ");
+
+        return new CommonPanacheQueryImpl<>(this, select.toString() + query, "select count(*) " + query);
+    }
+
+    public void filter(String filterName, Map<String, Object> parameters) {
+        if (filters == null)
+            filters = new HashMap<>();
+        filters.put(filterName, parameters);
+    }
+
+    public void page(Page page) {
+        this.page = page;
+        this.range = null; // reset the range to be able to switch from range to page
+    }
+
+    public void page(int pageIndex, int pageSize) {
+        page(Page.of(pageIndex, pageSize));
+    }
+
+    public void nextPage() {
+        checkPagination();
+        page(page.next());
+    }
+
+    public void previousPage() {
+        checkPagination();
+        page(page.previous());
+    }
+
+    public void firstPage() {
+        checkPagination();
+        page(page.first());
+    }
+
+    // FIXME: what do we do for this one?
+    //    public void lastPage() {
+    //        checkPagination();
+    //        page(page.index(pageCount() - 1));
+    //    }
+
+    public Uni<Boolean> hasNextPage() {
+        checkPagination();
+        return pageCount().map(pageCount -> page.index < (pageCount - 1));
+    }
+
+    public boolean hasPreviousPage() {
+        checkPagination();
+        return page.index > 0;
+    }
+
+    public Uni<Integer> pageCount() {
+        checkPagination();
+        return count().map(count -> {
+            if (count == 0)
+                return 1; // a single page of zero results
+            return (int) Math.ceil((double) count / (double) page.size);
+        });
+    }
+
+    public Page page() {
+        checkPagination();
+        return page;
+    }
+
+    private void checkPagination() {
+        // FIXME: turn into Uni
```

Comment:
Well, it also means we throw in situations where it'd be better to return a failed uni.

Reply:
Hum, in fact I did it this way because ... you did it this way in your first implementation of Panache Rx so do as you want I'll update the MongoDB part later if needed ;)
I agree that you don't expect reactive code to throw, but this particular case is a developement error (like a NPE) so I think you can throw